java-checks/src/main/java/org/sonar/java/checks/SillyEqualsCheck.java

```diff
@@ -0,0 +1,124 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import org.sonar.api.server.rule.RulesDefinition;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.methods.MethodInvocationMatcher;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.Tree.Kind;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.List;
+
+@Rule(
+  key = "S2159",
+  name = "Silly equality checks should not be made",
+  tags = {"bug", "unused"},
+  priority = Priority.MAJOR)
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.LOGIC_RELIABILITY)
+@SqaleConstantRemediation("15min")
+public class SillyEqualsCheck extends SubscriptionBaseVisitor {
+
+  private static final String JAVA_LANG_OBJECT = "java.lang.Object";
+
+  private static final String MESSAGE = "Remove this call to \"equals\"; comparisons between unrelated types always return false.";
+
+  private static final MethodInvocationMatcher EQUALS_METHOD = MethodInvocationMatcher.create()
+    .name("equals")
+    .addParameter(JAVA_LANG_OBJECT);
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodInvocationTree methodTree = (MethodInvocationTree) tree;
+    if (EQUALS_METHOD.matches(methodTree)) {
+      ExpressionTree firstArgument = Iterables.getOnlyElement(methodTree.arguments());
+      Type argumentType = firstArgument.symbolType().erasure();
+      Type ownerType = getMethodOwnerType(methodTree).erasure();
+      if (isLiteralNull(firstArgument)) {
+        addIssue(tree, "Remove this call to \"equals\"; comparisons against null always return false; consider using '== null' to check for nullity.");
+      } else if (ownerType.isArray()) {
+        checkWhenOwnerIsArray(tree, ownerType, argumentType);
+      } else {
+        checkWhenOwnerIsNotArray(tree, ownerType, argumentType);
+      }
+    }
+  }
+
+  private void checkWhenOwnerIsArray(Tree tree, Type ownerType, Type argumentType) {
+    if (!argumentType.isArray()) {
```

Comment:
Question of readability (and subject to debate) but I would rather check things without negation : if argument is an array... if arguments are unrelated etc. Negation adds difficulty for the reader.